lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/MainToolBar.kt

```diff
@@ -160,6 +142,69 @@ class MainToolBar(context: Context) : RelativeLayout(context) {
         clearComponent()
     }
 
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        super.onLayout(changed, l, t, r, b)
+        // Bypass regular layout for the titleComponent to make final placements based on the measurement
+        val titleComponent = getTitleComponent()
+        if (titleComponentAlignment == Alignment.Center) {
+            val titleLeft = max(l + leftButtonsBar.measuredWidth, (r - l) / 2 - titleComponent.measuredWidth / 2)
+            val titleRight = min(r - rightButtonsBar.measuredWidth, (r - l) / 2 + titleComponent.measuredWidth / 2)
+            titleComponent.layout(titleLeft, t, titleRight, b)
+        } else {
+            titleComponent.layout(l + leftButtonsBar.measuredWidth + defaultMargin, t, r - rightButtonsBar.measuredWidth, b)
+        }
+    }
+
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+        val mode = MeasureSpec.getMode(widthMeasureSpec)
+        val parentWidth = MeasureSpec.getSize(widthMeasureSpec)
+        val parentHeight = MeasureSpec.getSize(heightMeasureSpec)
+        logd("OnMeasure: parentWidth: $parentWidth, parentHeight: $parentHeight widthMeasureSpec:" +
+                " ${MeasureSpec.toString(widthMeasureSpec)}, heightMeasureSpec: ${MeasureSpec.toString(heightMeasureSpec)}", "MainToolBar")
+        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+        // exact measurement of the parent
+        if (mode == MeasureSpec.EXACTLY) {
+            val titleComponent = getTitleComponent()
+            // measure all 3 child with AT_MOST to get what space they require
+            rightButtonsBar.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+            leftButtonsBar.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+            titleComponent.measure(MeasureSpec.makeMeasureSpec(parentWidth, MeasureSpec.AT_MOST), heightMeasureSpec)
+
+            val titleComponentWidth = titleComponent.measuredWidth
+
+            // hide buttons title bypass the right end, otherwise show
+            var leftBarWidth = leftButtonsBar.measuredWidth
+            val titleLeft = parentWidth / 2 - (titleComponentWidth / 2)
+            if (titleLeft < this.left + leftBarWidth + defaultMargin) {
```

Comment:
Changing the `showAsAction` property from a function intended to measure views is wrong. This side effect is undesirable as it makes the code very difficult to reason about.

Reply:
Agree 100%, the decision about to shrink buttons or not is relying on the 3 measurements, maybe the place in `onMeasure` is not the right place, but it's the closest (fastest) place to get those measurements right, we can add the `doOnLayout` block to run after all these calculations are ready, that means, we should trigger that on title changes (text, alignment..), buttons changes ( add/remove/change text/ icon/ custom components).

 I thought about that, this is the closest position to what we need to achieve, WDYT about alternatives?

Reply:
Measuring the title bar and buttons bar should be enough for them to not overlap one another. Changing `showAsAction` should not be needed.